Add solve_pnp_dlt

thatbrguy opened the pull request:
#### Changes
This PR introduces the function `solve_pnp_dlt`. This function attempts to solve the Perspective-n-Point (PnP) problem using Direct Linear Transform (DLT).

##### Questions and Notes:
1. For the test function `test_project`, for dtype `float32` I have given easier tolerances (`atol` and `rtol` as `1e-1`). I think the error between ground truth image points and estimated image points is large when the estimated world to camera transformation matrix has some error and the camera coordinates are close to 0. For `float64` however stricter tolerances (`atol` and `rtol` as `1e-4`) seem to work. Do let me know if the easier tolerances are ok for `float32`. 
2. For `test_gradcheck` I had given `atol` as `1e-4` so that a test could pass. Do let me know if this fine as well.
3. I ran the following tests:
    - `pytest test/geometry/calibration/test_pnp.py --device all --dtype "float32,float64"`
    - `make lint`
    - `make mypy`
    - `make build-docs`
     
Do let me know if I have to make any changes, thanks!

#### Type of change
<!-- Please delete options that are not relevant. -->
- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update


#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings (**note:** I need to verify this, however I did run the tests I mentioned in point 3 of the Questions and Notes section above.)
- [ ] Did you update CHANGELOG in case of a major change?

thatbrguy commented:
I think the code has to be made compatible with the older versions of pytorch which kornia supports. Also I have to look into why some other tests failed. I'll look into them later tonight. Making this PR a draft for the time being.

edgarriba commented on kornia/geometry/calibration/pnp.py:
```diff
@@ -0,0 +1,131 @@
+    points lie on a line or if all the 3D world points lie on a plane.
+
+    Args:
+        world_points        :   A tensor with shape :math:`(N, 3)` representing
```
Should we provide batch mode ? BxNx3

edgarriba commented on kornia/geometry/calibration/pnp.py:
```diff
@@ -0,0 +1,131 @@
+    world_points_h = convert_points_to_homogeneous(world_points)
+
+    # Getting normalized image points.
+    norm_points = torch.matmul(intrinsic_inv, img_points_h.T).T
```
We have `transform_points`

edgarriba commented on test/geometry/calibration/test_pnp.py:
```diff
@@ -0,0 +1,106 @@
+
+    def _project_to_image(self, world_points, world_to_cam_4x4, intrinsic):
+        world_points_h = convert_points_to_homogeneous(world_points)
+        cam_points = (torch.matmul(world_to_cam_4x4, world_points_h.T).T)[:, :3]
```
Use transform_points

edgarriba commented on test/geometry/calibration/test_pnp.py:
```diff
@@ -0,0 +1,106 @@
+            shape=(num_points, 3), low=-100, high=100, dtype=dtype, device=device,
+        )
+
+        rotation = kornia.angle_axis_to_rotation_matrix(angle_axis)
```
```suggestion
        rotation = kornia.geometry.angle_axis_to_rotation_matrix(angle_axis)
```

edgarriba commented on test/geometry/calibration/test_pnp.py:
```diff
@@ -0,0 +1,106 @@
+        intrinsic, gt_world_to_cam, world_points, img_points = \
+            self._get_test_data(num_points, device, dtype)
+
+        pred_world_to_cam = kornia.solve_pnp_dlt(world_points, img_points, intrinsic)
```
```suggestion
        pred_world_to_cam = kornia.geometry.solve_pnp_dlt(world_points, img_points, intrinsic)
```

edgarriba commented:
> I think the code has to be made compatible with the older versions of pytorch which kornia supports. Also I have to look into why some other tests failed. I'll look into them later tonight. Making this PR a draft for the time being.

Probably we should make now the transition and deprecate pytorch <1.8
https://pepy.tech/project/Torch?versions=1.8.1&versions=1.9.0&versions=1.9.1&versions=1.6.0&versions=1.7.0&versions=1.7.1&versions=1.8.0

thatbrguy commented:
I have made this PR ready for review. @edgarriba I think I have addressed the suggestions you had made, do let me know if you have any more suggestions. 

I have made the tolerances for some of the tests more lenient (used `atol` as `1e-4` for `test_gradcheck` and used `atol`, `rtol` as `1e-3` for `test_pred_world_to_cam`). I have also introduced batch mode. I also changed operations so that the code may work with pytorch 1.6 (however I did not test with pytorch 1.6, let's see if the CI test passes). Let's see if the important tests pass.

I don't know if deprecating pytorch < 1.8 is a good idea though. Maybe the others can mention their opinions.

thatbrguy commented:
Hmm some important tests are failing still. I'll fix them later tonight. Making this PR draft until then.

edgarriba commented on test/geometry/calibration/test_pnp.py:
```diff
@@ -0,0 +1,137 @@
+            shape=(3,), low=-100, high=100, dtype=dtype, device=device,
+        )
+
+        rotation = kornia.geometry.angle_axis_to_rotation_matrix(angle_axis)
```
I think the best here is to create a random quaternion and convert to rotation_matrix to assure that is orthogonal and has determinant == +1
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.special_ortho_group.html#scipy.stats.special_ortho_group

edgarriba commented on test/geometry/calibration/test_pnp.py:
```diff
@@ -0,0 +1,137 @@
+        pred_world_to_cam = kornia.geometry.solve_pnp_dlt(world_points, img_points, intrinsics)
+        assert pred_world_to_cam.shape == (batch_size, 3, 4)
+
+    @pytest.mark.parametrize("num_points", (6, 20, 200))
```
could you check locally approx if the last test is "fast" ? usually gradcheck with large tensor slow down the testing

edgarriba commented on kornia/geometry/calibration/pnp.py:
```diff
@@ -0,0 +1,139 @@
+    norm_col = torch.sqrt(torch.sum(pred_world_to_cam[:, :3, 0] ** 2, axis=1))
+    mul_factor = (1 / norm_col)[:, None, None]
+    pred_world_to_cam = pred_world_to_cam * mul_factor
+
```
could you add a TODO note here for the refinenement step ?

thatbrguy replied:
I noticed that there are functions called `find_homography_dlt` and `find_homography_dlt_iterated`. Similarly, do you think having a separate `solve_pnp_dlt_iterated` instead of modifying `solve_pnp_dlt` would be nice ?

Also, do you think I should add a `NOTE` to the docstring of `solve_pnp_dlt` mentioning that the pose could be refined by using optimization or something? Or is just a TODO in the function to implement refinement in the future enough?

edgarriba replied:
add just todo

thatbrguy commented on kornia/geometry/calibration/pnp.py:
```diff
@@ -0,0 +1,139 @@
+
+    Returns:
+        A tensor with shape :math:`(B, 3, 4)` representing the estimated world to
+        camera transformation matrices.
```
The matrix that is being returned is the [R|t] matrix so it is actually the extrinsic matrix that is being returned. The projection matrix is K @ [R|t] right? So I think the suggestion should be `a.k.a. the extrinsic matrices`. I'll mention `also known as the extrinsic matrices` in my commit.

thatbrguy commented:
Made the PR ready for review. Looks like all tests except the add-emoji test has passed :tada:

lferraz commented:
> **EDIT:** If the final QR step is deemed to be unnecessary, I can just remove it for all versions also. Right now, all numeric tests pass with and without QR.

agree with you, probably it can be removed.

lferraz commented:
> ```
> >>> torch.abs(opencv_pred_world_to_cam - gt_world_to_cam).max() # Comparing OpenCV estimate and GT
> tensor(3.433228e-05)
> >>> 
> >>> torch.abs(pred_world_to_cam - gt_world_to_cam).max() # Comparing current PR's estimate and GT
> tensor(2.746582e-04)
> ```
> 
> 
>     
>   
> 
> Looks like the current PR's estimate is decent too. Also do note that OpenCV's approach uses LM to refine the estimate, whereas we do not. So I think once we add a refinement step (in a future PR) we may get better performance.

Congratulations, very good work!!!!!
Let's see if our LM can provide even better results than the one for OpenCV.
